src/hash_map.c

```diff
@@ -0,0 +1,545 @@
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifdef __cplusplus
+extern "C"
+{
+#endif
+
+#include <string.h>
+#include <stdio.h>
+
+#include "rcutils/allocator.h"
+#include "rcutils/error_handling.h"
+#include "rcutils/logging_macros.h"
+#include "rcutils/types/array_list.h"
+#include "rcutils/types/hash_map.h"
+#include "rcutils/types/rcutils_ret.h"
+#include "rcutils/macros.h"
+#include "rcutils/visibility_control.h"
+
+#define LOAD_FACTOR         (0.75)
+#define BUCKET_INITIAL_CAP  ((size_t)2)
+
+typedef struct rcutils_hash_map_entry_t
+{
+  size_t hashed_key;
+  void * key;
+  void * value;
+} rcutils_hash_map_entry_t;
+
+typedef struct rcutils_hash_map_impl_t
+{
+  // This is the array of buckets that will store the keypairs
+  rcutils_array_list_t * map;
+  size_t capacity;
+  size_t size;
+  size_t key_size;
+  size_t data_size;
+  rcutils_hash_map_key_hasher_t key_hashing_func;
+  rcutils_hash_map_key_cmp_t key_cmp_func;
+  rcutils_allocator_t allocator;
+} rcutils_hash_map_impl_t;
+
+// djb2 hash function
+size_t rcutils_hash_map_string_hash_func(const void * key_str)
+{
+  const char ** ckey_ptr = (const char **) key_str;
+  const char * ckey_str = *ckey_ptr;
+  size_t hash = 5381;
+
+  while ('\0' != *ckey_str) {
+    int c = *(ckey_str++);
+    hash = ((hash << 5) + hash) + c; /* hash * 33 + c */
+  }
+
+  return hash;
+}
+
+int rcutils_hash_map_string_cmp_func(const void * val1, const void * val2)
+{
+  const char ** cval1 = (const char **) val1;
+  const char ** cval2 = (const char **) val2;
+  return strcmp(*cval1, *cval2);
+}
+
+rcutils_hash_map_t
+rcutils_get_zero_initialized_hash_map()
+{
+  static rcutils_hash_map_t zero_initialized_hash_map = {NULL};
+  return zero_initialized_hash_map;
+}
+
+// Allocates a new zero initialized hashmap
+static rcutils_ret_t hash_map_allocate_new_map(
+  rcutils_array_list_t ** map, size_t capacity,
+  const rcutils_allocator_t * allocator)
+{
+  *map = allocator->allocate(capacity * sizeof(rcutils_hash_map_impl_t), allocator->state);
+  if (NULL == *map) {
+    return RCUTILS_RET_BAD_ALLOC;
+  }
+
+  // Make sure the list for every bucket is zero initialized
+  rcutils_array_list_t zero_list = rcutils_get_zero_initialized_array_list();
+  for (size_t i = 0; i < capacity; ++i) {
+    (*map)[i] = zero_list;
+  }
+
+  return RCUTILS_RET_OK;
+}
+
+// Deallocates the memory for a map entry
+static void hash_map_deallocate_entry(
+  rcutils_allocator_t * allocator,
+  rcutils_hash_map_entry_t * entry)
+{
+  if (NULL != entry) {
+    if (NULL != entry->key) {
```

Comment:
I think this condition can be removed since `deallocate(NULL, ...)` (ie `free(NULL)`) is a noop.

Same for `entry->value` below.

Reply:
Removed check on `entry->key` and `entry->value` being null before calling deallocate. Kept the check on `entry` so as to not dereference a null pointer.